cmd/src/dump.rs

```diff
@@ -0,0 +1,108 @@
+use engine_rocks::{
+    self,
+    raw::{DBOptions, Env},
+    RocksEngine,
+};
+use engine_traits::{Iterable, RaftEngine, RaftLogBatch, CF_DEFAULT};
+use kvproto::raft_serverpb::RaftLocalState;
+use protobuf::Message;
+use raft::eraftpb::Entry;
+use std::fs;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+/// Check the potential original raftdb directory and try to dump data out.
+///
+/// Procedure:
+///     1. Check whether the dump has been completed. If there is a flag file,
+///        delete the original raftdb safely and return.
+///     2. Scan and dump raft data into raft engine.
+///     3. Create a flag file to indicate that dump operation is done.
+///     4. Delete the original raftdb safely.
+pub fn check_and_dump_raft_db<E: RaftEngine>(path: &str, engine: &E, env: Arc<Env>) {
+    if !RocksEngine::exists(path) {
+        let flag_path = generate_flag_path(path);
+        let _ = fs::remove_file(flag_path);
```

Comment:
if the raftdb path doesn't exist, how could the flag file exist?

Reply:
For example, when last time the raftdb dir has been removed, but flag file isn't removed successfully due to some error like power off.